### PR TITLE
Print js-constructor function type names

### PR DESCRIPTION
**Pull request by sandersn**
Instead of printing them as a type literal, which is scary.

Also, print the names of otherwise-anonymous functions and classes as the name of the declaration that they are assigned to:

```ts
var o = {
  C: class { }
}
var oc = new o.C()
```

In both Typescript and Javascript, we currently print "(Anonymous class)" for the type of `oc`. After this change, we print it as `C`. This isn't perfect (the accessible symbol chain isn't there), but it's better than what we have.

### CODE REVIEW

File: tests/baselines/reference/exportNestedNamespaces.types [hunk 1]
```diff
@@ -7,13 +7,13 @@ exports.n = {};
 >{} : typeof __object
 
 exports.n.K = function () {
->exports.n.K = function () {    this.x = 10;} : () => void
->exports.n.K : () => void
+>exports.n.K = function () {    this.x = 10;} : typeof (Anonymous function)
```
Comment (weswigham) [inline tests/baselines/reference/exportNestedNamespaces.types]: Rather than `(Anonymous function)`, can we wire this up so it has name `K`?

Comment (sandersn) [reply tests/baselines/reference/exportNestedNamespaces.types]: Yep, I'm working on it. I'm trying to understand the machinery that decides between `K`, `n.K` and `exports.n.K`.

File: src/compiler/utilities.ts [hunk 2]
```diff
@@ -4581,6 +4587,23 @@ namespace ts {
...
     }
 
+    function getAssignedName(node: Node): DeclarationName {
+        if (!node.parent) {
+            return undefined;
+        }
+        else if (isPropertyAssignment(node.parent)) {
+            return node.parent.name;
+        }
+        else if (isBinaryExpression(node.parent) && node === node.parent.right) {
```
Comment (weswigham) [inline src/compiler/utilities.ts]: No love for binding pattern initializers?:
```ts
const {
a = function() {},
b = class {}
} = ({} as any);
```

**Comment by sandersn** [comment]
Looks like I forgot to update some fourslash tests. I'll do that and then add support for binding pattern initializers.

File: src/compiler/utilities.ts (hunk 2 above)
Comment (sandersn) [reply src/compiler/utilities.ts]: Done! (But, no, I don't like binding pattern initializers)

**Pull request merged**